src/problem.jl

```diff
@@ -1,13 +1,63 @@
 abstract type AbstractOptimizationProblem end
 
-struct OptimizationProblem{F,X,P,B,K} <: AbstractOptimizationProblem
+struct OptimizationProblem{F,X,P,B,LC,UC,K} <: AbstractOptimizationProblem
     f::F
     x::X
     p::P
     lb::B
     ub::B
+    lcons::LC
+    ucons::UC
     kwargs::K
-    function OptimizationProblem(f, x; p=DiffEqBase.NullParameters(), lb = nothing, ub = nothing, kwargs...)
-        new{typeof(f), typeof(x), typeof(p), typeof(lb), typeof(kwargs)}(f, x, p, lb, ub, kwargs)
+    function OptimizationProblem(f, x; p=DiffEqBase.NullParameters(), lb = [], ub = [], lcons = [], ucons = [], kwargs...)
+        new{typeof(f), typeof(x), typeof(p), typeof(lb), typeof(lcons), typeof(ucons), typeof(kwargs)}(f, x, p, lb, ub, lcons, ucons, kwargs)
     end
 end
+
+import MathOptInterface
+const MOI = MathOptInterface
+
+struct MOIOptimizationProblem <: MOI.AbstractNLPEvaluator
+    prob::OptimizationProblem
+end
+
+MOI.eval_objective(moiproblem::MOIOptimizationProblem, x) = moiproblem.prob.f isa OptimizationFunction ? moiproblem.prob.f.f(x) : moiproblem.prob.f(x)
+
+MOI.eval_constraint(moiproblem::MOIOptimizationProblem, g, x) = moiproblem.prob.f isa OptimizationFunction ? g .= moiproblem.prob.f.cons(x) : error("Use OptimizationFunction to pass in constraints function")
+
+MOI.eval_objective_gradient(moiproblem::MOIOptimizationProblem, G, x) = moiproblem.prob.f isa OptimizationFunction ? moiproblem.prob.f.grad(G, x) : error("Use OptimizationFunction to pass in gradient function")
+
+MOI.eval_hessian_lagrangian(moiproblem::MOIOptimizationProblem, H, x, σ, μ) = moiproblem.prob.f isa OptimizationFunction ? σ.* moiproblem.prob.f.hess(H, x) + μ .*moiproblem.prob.f.cons_h(H,x) : error("Use OptimizationFunction to pass in hessian function")
+
+MOI.eval_constraint_jacobian(moiproblem::MOIOptimizationProblem, J, x) = moiproblem.prob.f isa OptimizationFunction ? moiproblem.prob.f.cons_j(J, x) : error("Use OptimizationFunction to pass in jacobian function")
+
+function MOI.jacobian_structure(moiproblem::MOIOptimizationProblem)
+    return Tuple{Int64,Int64}[(1,i) for i in 1:length(moiproblem.prob.x)]
+end
+
+function MOI.hessian_lagrangian_structure(moiproblem::MOIOptimizationProblem)
+    return vcat([[(j,i) for i in 1:length(moiproblem.prob.x)] for j in 1:length(moiproblem.prob.x)]...)
+end
+
```

Comment:
We should add the bounds for the parameters as well https://travis-ci.org/github/SciML/GalacticOptim.jl/builds/722644404#L662, I think https://github.com/jump-dev/MathOptInterface.jl/blob/cdf87e9376a086fb80ea731cece453f5a92878ca/src/Test/nlp.jl#L136-L148 is how you would do it

Reply:
Can you say nothing more precise about the hessian_lagrangian_structure than that?